app/src/main/java/com/alphawallet/app/viewmodel/TokenFunctionViewModel.java

```diff
@@ -664,4 +721,65 @@ private Intent completeTransferIntent(Context ctx, Token token, List<BigInteger>
         intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
         return intent;
     }
+
+    public void getAsset(Token token, BigInteger tokenId)
+    {
+        openseaDisposable = openseaService.getAsset(token, tokenId)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(asset -> onAsset(asset), this::onAssetError);
+    }
+
+    private void onAssetError(Throwable throwable)
+    {
+        Timber.d(throwable);
+    }
+
+    private void onAsset(String asset)
+    {
+        try
+        {
+            JSONObject result = new JSONObject(asset);
+            if (result.has("collection") && result.has("traits"))
+            {
+                computeRarity(getCountFromJson(result), getTraitsFromJson(result));
+            }
+        }
+        catch (JSONException e)
+        {
+            Timber.e(e);
+        }
+    }
+
+    private ArrayList<Trait> getTraitsFromJson(JSONObject result) throws JSONException
+    {
+        return new Gson().fromJson(result.getString("traits"),
+                new TypeToken<ArrayList<Trait>>()
+                {
+                }.getType());
+    }
+
+    private long getCountFromJson(JSONObject result) throws JSONException
+    {
+        JSONObject collectionJson = result.getJSONObject("collection");
+        Stats stats = new Gson().fromJson(collectionJson.getString("stats"), Stats.class);
+        return stats.getCount();
+    }
+
+    private void computeRarity(long count, List<Trait> traitList) throws JSONException
+    {
+        for (Trait trait : traitList)
+        {
+            if (trait.getTraitCount() == 1)
+            {
+                trait.setTraitRarity(0); // Unique
+            }
+            else
+            {
+                trait.setTraitRarity(((float) trait.getTraitCount() * 100) / count);
```

Comment:
>Feature envy is a code smell describing when an object accesses fields of another object to execute some operation, instead of just telling the object what to do.

Can extract method `trait.updateRarity()`, it's much easier to test.